### DISCUSSION TITLE
Using Rector for checking namespace isolation

### DISCUSSION THREAD

**Discussion by dorrogeray** [rectorphp/rector#7110] [Ideas]
Hi,

I would like to create a tool which will enforce a rule where only classes from specified namespaces are allowed in some namespaces. Something along the lines of:

- `App\Services\<SomeNameSpace>\**`
  - `App\Services\<SomeNameSpace>\**`
  - `App\Models\<SomeNameSpace>\**`
  - `App\Exceptions\**`

This would mean that any class inside `App\Services\<SomeNameSpace>\**` namespace could only use classes from following namespaces (+vendor):

  - `App\Services\<SomeNameSpace>\**`
  - `App\Models\<SomeNameSpace>\**`
  - `App\Exceptions\**`

If a violation is found, it would be reported.

I like the power of Rector very much, but I am not sure if it makes sense to try to implement this as a rector rule. These violations will not be something that can be fixed automatically, only thing I can think of is Rector adding comments like `// WARNING: Usage of  class XXX in namespace YYY is not allowed`, but it seems a bit obnoxious.

What do you think, should I try to create a dedicated tool or is there a way to fit this rule into rector in some reasonable fashion?

Thanks for your feedback!

**Comment by TomasVotruba** [comment]
Hi, thanks for the question.
We use something similar for e.g.entity manager is allowed only in repository, or mapper is allowed only in repository etc.

Rector is rather A → B refacotring tool. What you look for is reporting tool - PHPStan.
The rule I mention above is here: https://github.com/symplify/phpstan-rules/blob/main/docs/rules_overview.md#exclusivedependencyrule

It does not specifically check the namespaces, but you can easily extend to handle that.

**Reply by dorrogeray** [reply]
Perfect, thanks, I will check it out!

**Reply by dorrogeray** [reply]
@TomasVotruba Followup PR is here: https://github.com/symplify/symplify/pull/4042

**Comment by gisostallenberg** [comment]
@dorrogeray With [deptrac](https://github.com/qossmic/deptrac) you can configure advanced rules for that and even generate an image of your dependencies and errors in the dependencies of one layer on the other layer.

**Reply by dorrogeray** [reply]
This looks really cool! Thanks for the suggestion! :+1:

**Reply by gisostallenberg** [reply]
I find the config can be hard, but the results are awesome. You'll find issues with dependencies you'd normally 'just add'.

**Reply by dorrogeray** [reply]
I have read the documentation, I will definitely give it a try. It will take some time to figure out the right rules, but it will be worth it, no doubt. I really like the "categorization" engine. I wonder if it could be extracted into a reusable package, so other tools like phpstan / rector can leverage it, to allow applying some controversial / specific rules based on "layers". Rector's skipper does something similar, but this would be even more powerful.
